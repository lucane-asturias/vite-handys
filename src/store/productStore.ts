import { defineStore } from 'pinia'
import type { Product } from '@/interfaces'
const BASE_URL = 'http://faceprog.ru/reactcourseapi/products'

export const useProductStore = defineStore('productStore', {
	state: () => ({
		products: []
	}) as Products[],
	actions: {
		setProducts(products: Product) {
			this.products = products
		},
		async loadProducts() {
			const response = await fetch(`${BASE_URL}/all.php`)
			const products = await response.json() as Product[]

			this.setProducts(products)
		}
	},
	getters: {
		getAllProducts: state => state.products,
		findProductById: state => (id: string) => 
      state.products?.find(product => product.id === Number(id))
	},
})