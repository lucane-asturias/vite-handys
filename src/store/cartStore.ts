import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { useProductStore } from '@/store/productStore'
import type { Product, Cart, CartToken } from '@/interfaces'
const BASE_URL = 'http://faceprog.ru/reactcourseapi/cart/'

export const useCartStore = defineStore('cartStore', () => {

  // State Properties ============================

	const carts = ref<Cart[] | null>([])
	const token = ref<string | null>(null)

	// Actions ============================

  function setCart(cart: Cart, tkn: string) {
    carts.value = cart
    token.value = tkn
  }

	async function loadCart() {
		const oldToken = localStorage.getItem('CART__TOKEN') || null
		const response = await fetch(`${BASE_URL}load.php?token=${oldToken}`)
    // if token does not exists, the api generates one
		const { cart, token: tkn, needUpdate } = await response.json() as CartToken

		if (needUpdate) localStorage.setItem('CART__TOKEN', tkn)

    const formattedCart = cart.map(c => ({ id: c.id, qnt: c.cnt }))

		setCart(formattedCart, tkn)
	}

  async function addCart(id: number) {
		if (!inCart(id)) {
			const response = await fetch(
        `${BASE_URL}add.php?token=${token.value}&id=${id}`
      )

			const result: boolean = await response.json()

			if (result) {
        carts.value?.push({ id, qnt: 1 })
      }
		}
  }
	
  async function removeCart(id: number) {
		if (inCart(id)) {
			const response = await fetch(
        `${BASE_URL}remove.php?token=${token.value}&id=${id}`
      )
			const result = await response.json()

			if (result && carts.value) {
				carts.value = carts.value.filter(cart => cart.id != id)
			}
		}
  }

  function setQnt({ id, qnt }: { id: number, qnt: number }) {
		if (inCart(id)) {
			const cartDetail: Product = cartsDetailed().find(cart => cart.id == id)
			const validQnt = Math.min(Math.max(qnt, 1), cartDetail.rest)

      carts.value ? carts.value.find(cart => cart.id == id).qnt = validQnt : null
		}
	}

  // Getters ============================

  function inCart(id: number) {
    return carts.value?.some(cart => cart.id === id)
  }

  function cartsDetailed(): Product[] {
    const productStore = useProductStore()
    // âœ¨ Get all carts associated with a product by it's id 
    return carts.value?.map(cart => {
      let product = productStore.findProductById(cart.id)
      return { ...product, qnt: cart.qnt }
    })
  }

  const cartsLength = computed(() => carts.value?.length)

  const cartsTotal = computed<number>(() => {
    if (cartsDetailed().length > 0) {
      return cartsDetailed().reduce((t, pr) => t + pr.price * pr.qnt, 0)
    }
    return 0
  })

  return { 
    carts, token, loadCart, 
    addCart, removeCart, setQnt, inCart,
    cartsDetailed, cartsLength, cartsTotal
  }

})
