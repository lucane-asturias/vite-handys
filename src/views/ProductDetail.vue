<script lang="ts" setup>
  import { computed, defineAsyncComponent } from 'vue'
	import { useProductStore } from '@/store/productStore'
  import formatPrice from '@/helpers/formatPrice'

  const AppE404 = defineAsyncComponent(() => import('@/views/E404.vue'))

  const props = defineProps<{ id: string }>() // route segment id

  const productStore = useProductStore()
  const product = computed(() => productStore.findProductById(props.id) ?? null)
</script>
<template>
	<div class="mt-5" v-if="product">
		<p class="text-3xl font-bold">{{ product.title }}</p>
		<hr class="h-px my-2">
		<div class="alert p-4 rounded-lg border-1 border-emerald-700">
			Price: {{ formatPrice(product.price) }}
		</div>
    <router-link :to="{ name: 'catalog' }" class="mt-4 inline-block rounded py-2 px-3 bg-emerald-600 text-white hover:bg-emerald-700">
    	Back to catalog
      </router-link>
	</div>
	<app-e-404 class="mt-5" v-else />
</template>

<style lang="css" scoped>
.alert {
  background-color: #d1e7dd;
  color: #0a362;
  border-color: #a3cfbb
}
</style>