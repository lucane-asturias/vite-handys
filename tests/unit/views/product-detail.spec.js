import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia'
import { createTestingPinia } from '@pinia/testing'
import ProductDetail from '@/views/ProductDetail.vue'

import { productsMockData } from '../__mocks__/mock-data/products-data'
import { useProductStore } from '@/store/productStore'

describe('ProductDetail.vue component', () => {
  var formatPrice = vi.fn(num => num)
  var mockRouter = { push: vi.fn() }

  let wrapper
  let productStore
  // SETUP - run prior to each unit test
  beforeEach(() => {
    setActivePinia(createPinia())

    productStore = useProductStore()

    wrapper = shallowMount(ProductDetail, {
      props: { id: '101' },
      global: { 
        components: { 'router-link': RouterLinkStub },
        plugins: [createTestingPinia({
          initialState: {
            productStore: { products: productsMockData }
          }
        })]
      },
    })
  })

  afterEach(() => wrapper.unmount())

  test('should match snapshot', () => expect(wrapper.html()).toMatchSnapshot())

  test('should render product info correctly', () => {
    expect(wrapper.find('p').text()).toBe('Samsung AAZ8')
    expect(wrapper.find('.alert').text()).toBe('Price: $22,000.00')
  })

  test('to prop should push to catalog page whenever router-link is clicked', () => {
    const expectedObj = { name: 'catalog' }
    expect(wrapper.findComponent(RouterLinkStub).props().to).toMatchObject(expectedObj)
  })

})