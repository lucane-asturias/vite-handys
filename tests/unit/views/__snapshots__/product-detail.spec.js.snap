// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductDetail.vue component > should match snapshot 1`] = `
"<div data-v-95e08fb6=\\"\\" class=\\"mt-5\\">
  <p data-v-95e08fb6=\\"\\" class=\\"text-3xl font-bold\\">Samsung AAZ8</p>
  <hr data-v-95e08fb6=\\"\\" class=\\"h-px my-2\\">
  <div data-v-95e08fb6=\\"\\" class=\\"alert p-4 rounded-lg border-1 border-emerald-700\\"> Price: $22,000.00</div>
  <router-link-stub data-v-95e08fb6=\\"\\" to=\\"[object Object]\\" custom=\\"false\\" class=\\"mt-4 inline-block rounded py-2 px-3 bg-emerald-600 text-white hover:bg-emerald-700\\"></router-link-stub>
</div>"
`;
